NetKet/GroundState/variational_montecarlo.hpp

```diff
@@ -64,25 +63,18 @@ class VariationalMonteCarlo {
   Eigen::MatrixXd vsamp_;
 
   Eigen::VectorXcd grad_;
-  Eigen::VectorXcd gradprev_;
-
-  double sr_diag_shift_;
-  bool sr_rescale_shift_;
-  bool use_iterative_;
 
   int totalnodes_;
   int mynode_;
 
   AbstractOptimizer &opt_;
+  SR sr_;
+  bool dosr_;
```

Comment:
I have been thinking about whether we can get rid of the `dosr_` flag (which at the same time would make the code more modular).

One option would be to make a small class hierarchy like this (just pseudo-code):
```
class Updater {
public:
  void ComputeUpdate(...) = 0;
};

class GradientBasedUpdater : public Updater {
  Optimizer opt_;
public:
  override void ComputeUpdate(...) {
    // Compute deltap using opt.Update
  }
};

 // The current SR class with only the code for the holomorphic update
class SRHolomorphic  : public Updater {
public:
  override void ComputeUpdate(...) {
    // ...
  }
};

 // The current SR class with only the code for the non-holomorphic update
class SRReal  : public Updater {
public:
  override void ComputeUpdate(...) {
    // ...
  }
};
```

This has the additional benefit that a) we could more easily incorporate strategies that are neither SR nor `Optimizer`-like and b) it more cleanly separates the holomorphic and non-holomorphic SR code, which should make the code easier to understand.

Reply:
Good idea @femtobit , the remaining issue is really defining an interface that makes sense/is general enough for `ComputeUpdate` in `Updater` then. I agree that this structure would eliminate a lot of ugly and redundant code though....